File: src/WebAPI.js
import request from "superagent/lib/client"
import saPrefix from "superagent-prefix"
import toNumber from "lodash/toNumber"
import {TILEHOSTING_API_KEY} from "./constants/App"
import {OFDB_API, TH_GEOCODER, NOMINATIM, CORS_PROXY, PROMINENT_TAGS, PUBLIC_RESOURCES} from "./constants/URLs"
import {NUMBER_TAGS_TO_FETCH} from "./constants/Search"
import parse from 'csv-parse/lib/sync'


const prefix = saPrefix(OFDB_API.link)
const publicResources = saPrefix(PUBLIC_RESOURCES.link)
const FALANSTER_TOKEN = 'eyJzdWIiOiJtYXBhLWZhbGFuc3RlciIsIm5hbWUiOiJmYWxhbn'

const jsonCallback = (cb) => (err, res) => {
  if (err) {
    cb(err)
  } else {
    cb(null, res.body)
  }
}

function normalizeCoordinate(bbox, idx) {
  if (bbox.length > idx && bbox[idx] && (!isNaN(bbox[idx])) && bbox[idx] > 180) {
    bbox[idx] = ((bbox[idx] + 180.0) % 360.0) - 180.0
  }
  if (bbox.length > idx && bbox[idx] && (!isNaN(bbox[idx])) && bbox[idx] < -180) {
    bbox[idx] = ((bbox[idx] - 180.0) % 360.0) + 180.0
  }
}

const transformRecordToOption = (record) => {
  const option = {
    ...record,
    styles: {
      bold: record.bold === "true",
      underline: record.underline === "true",
      italic: record.italic === "true",
      fontSize: toNumber(record.fontSize) || 20
    }
  }

  const keysToDelete = ['bold', 'underline', 'italic', 'fontSize']
  keysToDelete.forEach(key => delete option[key])

  return option
}

const transformCSVToOptions = (records) => {
  const options = []
  for (let i = 0; i < records.length; i++) {
    const record = records[i]
    if (record.headline === "true") {
      const option = {
        label: record.label,
        options: []
      }
      i++
      for (; i !== records.length; i++) {
        const subRecord = records[i]
        if (subRecord.headline === "true") {
          i--
          break
        }

        option.options.push(transformRecordToOption(subRecord))
      }

      options.push(option)
      continue
    }

    options.push(transformRecordToOption(record))
  }

  return options
}

module.exports = {

  searchEntries: (txt, cats, bbox, cb) => {

    if (txt == null) {
      txt = ''
    }
    if (cats == null) {
      cats = []
    }
    if (bbox == null) {
      bbox = []
    }
    normalizeCoordinate(bbox, 1)
    normalizeCoordinate(bbox, 3)
    request
      .get('/search')
      .use(prefix)
      .query({
        text: txt.trim()
      })
      .query((cats.length > 0) ? ('categories=' + cats.join(',')) : "")
      .query('bbox=' + bbox.join(','))
      .set('Accept', 'application/json')
      .end(jsonCallback(cb))
  },

  searchEvents: (txt, bbox, start, end, cb) => {
    if (bbox == null) {
      bbox = []
    }
    normalizeCoordinate(bbox, 1)
    normalizeCoordinate(bbox, 3)
    let req = request
      .get('/events')
      .use(prefix)
      .set('Accept', 'application/json')
    if (bbox && bbox.length > 0) req.query('bbox=' + bbox.join(','))
    if (txt && txt.length > 0) req.query({text: txt.trim()})
    if (start) req.query(start ? ('start_min=' + start) : "")
    if (end) req.query(end ? ('start_max=' + end) : "")

    req.end(jsonCallback(cb))
  },

  createNewEvent: (newEvent, callBack) => {
    request
      .post('/events')
      .use(prefix)
      .set({'Accept': 'application/json', 'Authorization': `Bearer ${FALANSTER_TOKEN}`})
      .send(newEvent)
      .end((err, res) => {
        if (err) {
          callBack(err)
        } else {
          callBack(null, res.text.replace(/"/g, ""))
        }
      })
  },

  editEvent: (event, callBack) => {
    request
      .put('/events/' + event.id)
      .use(prefix)
      .set({'Accept': 'application/json', 'Authorization': `Bearer ${FALANSTER_TOKEN}`})
      .send(event)
      .end((err, res) => {
        if (err) {
          callBack(err)
        } else {
          callBack(null, res.text)
        }
      })
  },

  getEvent: (ids = [], cb) => {
    if (!Array.isArray(ids)) {
      ids = [ids]
    }

    if (ids.length < 1) {
      cb(new Error("no IDs were passed"))
    } else {
      request
        .get('/events/' + ids.join(','))
        .use(prefix)
        .set('Accept', 'application/json')
        .end(jsonCallback(cb))
    }
  },

  countTags: (cb) => {
    request
      .get('/count/tags')
      .use(saPrefix(OFDB_API.link))
      .set('Accept', 'application/json')
      .end(jsonCallback(cb))
  },

  getProminentTags: (cb) => {
    const xhr = new XMLHttpRequest()
    xhr.open("GET", `${CORS_PROXY.link}/${PROMINENT_TAGS.link}`)

    xhr.onreadystatechange = function () {
      if (xhr.readyState === XMLHttpRequest.DONE) {
        var status = xhr.status
        if (status === 0 || (status >= 200 && status < 400)) {
          // The request has been completed successfully
          const el = document.createElement('html')
          el.innerHTML = xhr.responseText

          const prominentTags = el.querySelector("#post-6714 > div > p").textContent
            .split('#').filter(t => t.length)

          cb(prominentTags)
        } else {
          // Oh no! There has been an error with the request!
          cb([], "could not fetch prominent tags")
        }
      }
    }

    xhr.send()
  },

  getTags: (page, cb) => {
    const offset = page * NUMBER_TAGS_TO_FETCH
    const limit = NUMBER_TAGS_TO_FETCH

    request
      .get('/entries/most-popular-tags')
      .use(saPrefix(OFDB_API.link))
      .query({
        'min-count': 1,
        offset,
        limit,
      })
      .set('Accept', 'application/json')
      .end(jsonCallback(cb))
  },

  searchAddressTilehosting: (addr, cb) => {
    let query = TH_GEOCODER.link.replace("<query>", addr).replace("<key>", TILEHOSTING_API_KEY)
    if (addr != null && addr != "") {
      request
        .get(query)
        .set('Accept', 'application/json')
        .end(jsonCallback(cb))
    }
  },

  searchAddressNominatim: (addr, cb) => {
    // console.trace()
    if (addr == null) {
      addr = ''
    }
    request
      .get('/search')
      .use(saPrefix(NOMINATIM.link))
      .query({
        q: addr
      })
      .query({
        format: 'json'
      })
      .query({
        addressdetails: 1
      })
      .set('Accept', 'application/json')
      .end(jsonCallback(cb))
  },

  searchGeolocation: (latlng, cb) => {

    if (latlng == null) {
      latlng = {
        lat: 0.0,
        lng: 0.0
      }
    }

    request
      .get('/reverse')
      .use(saPrefix(NOMINATIM.link))
      .query({
        lat: latlng.lat
      })
      .query({
        lon: latlng.lng
      })
      .query({
        zoom: 18
      })
      .query({
        format: 'json'
      })
      .query({
        addressdetails: 1
      })
      .set('Accept', 'application/json')
      .end(jsonCallback(cb))
  },

  getEntries: (ids = [], cb) => {

    if (!Array.isArray(ids)) {
      ids = [ids]
    }

    if (ids.length < 1) {
      cb(new Error("no IDs were passed"))
    } else {
      request
        .get('/entries/' + ids.join(','))
        .use(prefix).set('Accept', 'application/json')
        .end(jsonCallback(cb))
    }
  },

  getRatings: (ids = [], cb) => {

    if (!Array.isArray(ids)) {
      ids = [ids]
    }

    if (ids.length < 1) {
      cb(new Error("no IDs were passed"))
    } else {
      request
        .get('/ratings/' + ids.join(','))
        .use(prefix).set('Accept', 'application/json')
        .end(jsonCallback(cb))
    }
  },

  saveNewEntry: (e, cb) => {
    request
      .post('/entries/')
      .use(prefix)
      .set('Accept', 'application/json')
      .send(e)
      .end((err, res) => {
        if (err) {
          cb(err)
        } else {
          cb(null, res.text.replace(/"/g, ""))
        }
      })
  },

  saveEntry: (e, cb) => {
    request
      .put('/entries/' + e.id)
      .use(prefix)
      .set('Accept', 'application/json')
      .send(e)
      .end((err, res) => {
        if (err) {
          cb(err)
        } else {
          cb(null, res.text)
        }
      })
  },

  createRating: (r, cb) => {
    request
      .post('/ratings/')
      .use(prefix)
      .set('Accept', 'application/json')
      .send(r)
      .end((err, res) => {
        if (err) {
          cb(err)
        } else {
          cb(null, res.text)
        }
      })
  },

  getAllCategories: (cb) => {
    request
      .get('/categories/')
      .use(prefix)
      .set('Accept', 'application/json')
      .end(cb)
  },

  getServerInfo: (cb) => {
    request
      .get('/server/version')
      .set('Accept', 'application/json')
      .use(prefix)
      .end((err, res) => {
        if (err) {
          cb(err)
        } else {
          cb(null, {
            version: res.text
          })
        }
      })
  },

  register: ({
               email,
               password,
             }, cb) => {
    request
      .post('/users')
      .use(prefix)
      .set('Accept', 'application/json')
      .send({
        email,
        password
      })
      .end(cb)
  },

  login: ({
            email,
            password
          }, cb) => {
    request
      .post('/login')
      .set('Accept', 'application/json')
      .use(prefix)
      .withCredentials()
      .send({
        email,
        password
      })
      .end(cb)
  },

  getUser: (email, cb) => {
    request
      .get('/users/' + email)
      .set('Accept', 'application/json')
      .use(prefix)
      .withCredentials()
      .end(cb)
  },

  logout: (cb) => {
    request
      .post('/logout')
      .set('Accept', 'application/json')
      .use(prefix)
      .withCredentials()
      .end(cb)
  },

  confirmEmail: (token, cb) => {
    request
      .post('/confirm-email-address')
      .set('Accept', 'application/json')
      .use(prefix)
      .send({
        token
      })
      .end(cb)
  },

  deleteAccount: (email, cb) => {
    request
      .delete('/users/' + email)
      .set('Accept', 'application/json')
      .use(prefix)
      .withCredentials()
      .end(cb)
  },

  subscribeToBbox: (bbox, cb) => {
    let coordinates = [bbox._southWest, bbox._northEast]
    request
      .post('/subscribe-to-bbox')
      .use(prefix)
      .set('Accept', 'application/json')
      .send(coordinates)
      .end((err, res) => {
        if (err) {
          cb(err)
        } else {
          cb(null, res.text)
        }
      })
  },

  getBboxSubscriptions: (cb) => {
    request
      .get('/bbox-subscriptions')
      .set('Accept', 'application/json')
      .use(prefix)
      .end(cb)
  },

  unsubscribeFromBboxes: (cb) => {
    request
      .delete('/unsubscribe-all-bboxes')
      .set('Accept', 'application/json')
      .use(prefix)
      .end(cb)
  },

  getDropdowns: (name, cb) => {
    const dropdowns = {
      'categories': [],
      'regions': [],
    }


    Promise.all(Object.keys(dropdowns).map(async (dropdownName) => {
      try {
        const res = await request
          .get(`/customizations/${name}/filters/${dropdownName}.csv`)
          .type('text/csv')
          .use(publicResources)

        const records = parse(
          res.text, {
            columns: true,
            skip_empty_lines: true,
          }
        )

        const transformed = transformCSVToOptions(records)
        dropdowns[dropdownName] = transformed
      } catch (e) {
        console.error(e)
      }
    })).then(cb(dropdowns))
  }
}

File: src/WebAPI.rs
use crate::{
    constants::{
        App::{self, TILEHOSTING_API_KEY},
        Categories::IDS,
        Map::DEFAULT_BBOX,
        URLs::{NOMINATIM, OFDB_API, TH_GEOCODER},
    },
    entities::*,
    Actions, Msg,
};
use futures::Future;
use seed::fetch::Request;

fn prefix() -> String {
    OFDB_API(&App::APP_STAGES::PRODUCTION)
}

// TODO: function normalizeCoordinate(bbox, idx) {
// TODO:     if (bbox.length > idx && bbox[idx] && (!isNaN(bbox[idx])) && bbox[idx] > 180) {
// TODO:         bbox[idx] = ((bbox[idx] + 180.0) % 360.0) - 180.0;
// TODO:     }
// TODO:     if (bbox.length > idx && bbox[idx] && (!isNaN(bbox[idx])) && bbox[idx] < -180) {
// TODO:         bbox[idx] = ((bbox[idx] - 180.0) % 360.0) + 180.0;
// TODO:     }
// TODO: }

pub fn searchEntries(
    txt: &Option<String>,
    cats: &[IDS],
    bbox: Option<&BBox>,
) -> impl Future<Item = Msg, Error = Msg> {
    // TODO: normalizeCoordinate(bbox, 1);
    // TODO: normalizeCoordinate(bbox, 3);

    let bbox: String = bbox_to_slice(bbox.unwrap_or(&default_bbox()))
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<_>>()
        .join(",");

    // TODO: .query((cats.length > 0) ? ('categories=' + cats.join(',')) : "")
    let cats: String = cats
        .into_iter()
        .map(|x| x.as_str())
        .collect::<Vec<_>>()
        .join(",");

    let url = format!(
        "{}/search?text={}&categories={}&bbox={}",
        prefix(),
        txt.as_ref().unwrap_or(&"".to_string()).trim(),
        cats,
        bbox
    );
    Request::new(url)
        .fetch_json_data(|d| Msg::Server(Actions::server::Msg::SEARCH_RESULT_ENTRIES(d)))
}

// TODO:   searchEvents: (tags, bbox, start, end, cb) => {
// TODO:     if (bbox == null) {
// TODO:       bbox = [];
// TODO:     }
// TODO:     normalizeCoordinate(bbox, 1);
// TODO:     normalizeCoordinate(bbox, 3);
// TODO:     let req = request
// TODO:       .get('/events')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json');
// TODO:     if(bbox && bbox.length > 0) req.query('bbox=' + bbox.join(','))
// TODO:     if(tags && tags.length > 0) req.query('tags=' + tags) // TODO
// TODO:     if(start) req.query(start ? ('start_min=' + start) : "")
// TODO:     if(end) req.query(end ? ('start_max=' + end) : "")
// TODO:
// TODO:     req.end(jsonCallback(cb));
// TODO:   },
// TODO:
// TODO:   getEvent: (id, cb) => {
// TODO:     request
// TODO:       .get('/events/' + id)
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .end(jsonCallback(cb));
// TODO:   },
// TODO:
// TODO:   searchAddressTilehosting: (addr, cb) => {
// TODO:     let query = TH_GEOCODER.link.replace("<query>", addr).replace("<key>", TILEHOSTING_API_KEY);
// TODO:     if (addr != null && addr != "") {
// TODO:       request
// TODO:         .get(query)
// TODO:         .set('Accept', 'application/json')
// TODO:         .end(jsonCallback(cb));
// TODO:     }
// TODO:   },

pub fn searchAddressNominatim(addr: &str) -> impl Future<Item = Msg, Error = Msg> {
    let url = format!(
        "{}/search?q={}&format=json&addressdetails=1",
        NOMINATIM::link,
        addr
    );
    Request::new(url)
        .fetch_json_data(|d| Msg::Server(Actions::server::Msg::SEARCH_ADDRESS_RESULT(d)))
}

// TODO:   searchGeolocation: (latlng, cb) => {
// TODO:
// TODO:     if (latlng == null) {
// TODO:       latlng = {
// TODO:         lat: 0.0,
// TODO:         lng: 0.0
// TODO:       };
// TODO:     }
// TODO:
// TODO:     request
// TODO:       .get('/reverse')
// TODO:       .use(saPrefix(NOMINATIM.link))
// TODO:       .query({
// TODO:         lat: latlng.lat
// TODO:       })
// TODO:       .query({
// TODO:         lon: latlng.lng
// TODO:       })
// TODO:       .query({
// TODO:         zoom: 18
// TODO:       })
// TODO:       .query({
// TODO:         format: 'json'
// TODO:       })
// TODO:       .query({
// TODO:         addressdetails: 1
// TODO:       })
// TODO:       .set('Accept', 'application/json')
// TODO:       .end(jsonCallback(cb));
// TODO:   },
// TODO:
// TODO:   getEntries: (ids = [], cb) => {
// TODO:
// TODO:     if (!Array.isArray(ids)) {
// TODO:       ids = [ids];
// TODO:     }
// TODO:
// TODO:     if (ids.length < 1) {
// TODO:       cb(new Error("no IDs were passed"));
// TODO:     } else {
// TODO:       request
// TODO:         .get('/entries/' + ids.join(','))
// TODO:         .use(prefix).set('Accept', 'application/json')
// TODO:         .end(jsonCallback(cb));
// TODO:     }
// TODO:   },
// TODO:
// TODO:   getRatings: (ids = [], cb) => {
// TODO:
// TODO:     if (!Array.isArray(ids)) {
// TODO:       ids = [ids];
// TODO:     }
// TODO:
// TODO:     if (ids.length < 1) {
// TODO:       cb(new Error("no IDs were passed"));
// TODO:     } else {
// TODO:       request
// TODO:         .get('/ratings/' + ids.join(','))
// TODO:         .use(prefix).set('Accept', 'application/json')
// TODO:         .end(jsonCallback(cb));
// TODO:     }
// TODO:   },
// TODO:
// TODO:   saveNewEntry: (e, cb) => {
// TODO:     request
// TODO:       .post('/entries/')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .send(e)
// TODO:       .end((err, res) => {
// TODO:         if (err) {
// TODO:           cb(err);
// TODO:         } else {
// TODO:           cb(null, res.text.replace( /"/g ,""));
// TODO:         }
// TODO:       });
// TODO:   },
// TODO:
// TODO:   saveEntry: (e, cb) => {
// TODO:     request
// TODO:       .put('/entries/' + e.id)
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .send(e)
// TODO:       .end((err, res) => {
// TODO:         if (err) {
// TODO:           cb(err);
// TODO:         } else {
// TODO:           cb(null, res.text);
// TODO:         }
// TODO:       });
// TODO:   },
// TODO:
// TODO:   createRating: (r, cb) => {
// TODO:     request
// TODO:       .post('/ratings/')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .send(r)
// TODO:       .end((err, res) => {
// TODO:         if (err) {
// TODO:           cb(err);
// TODO:         } else {
// TODO:           cb(null, res.text);
// TODO:         }
// TODO:       });
// TODO:   },
// TODO:
// TODO:   getAllCategories: (cb) => {
// TODO:     request
// TODO:       .get('/categories/')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   getServerInfo: (cb) => {
// TODO:     request
// TODO:       .get('/server/version')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .end((err, res) => {
// TODO:         if (err) {
// TODO:           cb(err);
// TODO:         } else {
// TODO:           cb(null, {
// TODO:             version: res.text
// TODO:           });
// TODO:         }
// TODO:       });
// TODO:   },
// TODO:
// TODO:   register: ({
// TODO:     email,
// TODO:     password,
// TODO:   }, cb) => {
// TODO:     request
// TODO:       .post('/users')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .send({
// TODO:         email,
// TODO:         password
// TODO:       })
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   login: ({
// TODO:     email,
// TODO:     password
// TODO:   }, cb) => {
// TODO:     request
// TODO:       .post('/login')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .withCredentials()
// TODO:       .send({
// TODO:         email,
// TODO:         password
// TODO:       })
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   getUser: (email, cb) => {
// TODO:     request
// TODO:       .get('/users/' + email)
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .withCredentials()
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   logout: (cb) => {
// TODO:     request
// TODO:       .post('/logout')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .withCredentials()
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   confirmEmail: (token, cb) => {
// TODO:     request
// TODO:       .post('/confirm-email-address')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .send({
// TODO:         token
// TODO:       })
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   deleteAccount: (email, cb) => {
// TODO:     request
// TODO:       .delete('/users/' + email)
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .withCredentials()
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   subscribeToBbox: (bbox, cb) => {
// TODO:     let coordinates = [bbox._southWest, bbox._northEast];
// TODO:     request
// TODO:       .post('/subscribe-to-bbox')
// TODO:       .use(prefix)
// TODO:       .set('Accept', 'application/json')
// TODO:       .send(coordinates)
// TODO:       .end((err, res) => {
// TODO:         if (err) {
// TODO:           cb(err);
// TODO:         } else {
// TODO:           cb(null, res.text);
// TODO:         }
// TODO:       });
// TODO:   },
// TODO:
// TODO:   getBboxSubscriptions: (cb) => {
// TODO:     request
// TODO:       .get('/bbox-subscriptions')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .end(cb);
// TODO:   },
// TODO:
// TODO:   unsubscribeFromBboxes: (cb) => {
// TODO:     request
// TODO:       .delete('/unsubscribe-all-bboxes')
// TODO:       .set('Accept', 'application/json')
// TODO:       .use(prefix)
// TODO:       .end(cb);
// TODO:   }
// TODO: };

const fn bbox_to_slice(bbox: &BBox) -> [f64; 4] {
    [
        bbox.south_west.lat,
        bbox.south_west.lng,
        bbox.north_east.lat,
        bbox.north_east.lng,
    ]
}

const fn default_bbox() -> BBox {
    use DEFAULT_BBOX::*;
    BBox {
        north_east: Coordinate {
            lat: _northEast::lat,
            lng: _northEast::lng,
        },
        south_west: Coordinate {
            lat: _southWest::lat,
            lng: _southWest::lng,
        },
    }
}
